drv/lpc55-spi-server/src/main.rs

```diff
@@ -8,24 +8,329 @@
+    // Last byte has been clocked out to SPI bus, return to Idle.
+    Error = 0xff, // Error during transmit.
+}
+
+impl From<u8> for TxState {
+    fn from(txstate: u8) -> Self {
+        match txstate {
+            0 => TxState::Idle,
+            1 => TxState::Queued,
+            2 => TxState::Writing,
+            3 => TxState::Finish,
+            _ => TxState::Error,
+        }
+    }
+}
+
 #[derive(Copy, Clone, PartialEq)]
 enum Trace {
-    IRQ,
-    Tx(u8),
-    Rx(u8),
+    AlreadyEnqueuedMsg,
+    BadMsgLength,
+    CannotSendRxFragError,
+    CollectPayload,
+    AssertRotIrq,
+    DeassertRotIrq,
+    Enqueued,
+    EnqueuedMsg(usize),
+    EnqueueFail,
+    EnqueueTooBig,
+    GetThePayload(usize),
+    HeaderCollect,
+    HeaderComplete,
+    HeaderOnly,
+    Initialized,
+    InvalidRxState(RxState),
+    Irq(bool, bool, bool),
+    Line,
+    SSAsserted,
+    SSDeasserted,
+    Loop(u16, u16, TxState, RxState),
+    LStat(bool, bool, bool, bool, bool, bool, bool, u8, u8),
+    PayloadComplete,
+    ResetRxToIdle,
+    Responding,
+    RespondToEcho,
+    RespondToStatus,
+    RespondToUnknown,
+    RxFragment,
+    Rx(RxState, u8, usize, bool, bool),
+    RxState(RxState),
+    Tx(TxState, u8, usize),
+    StartTx(bool, bool),
     None,
 }
 
-ringbuf!(Trace, 64, Trace::None);
+ringbuf!(Trace, 128, Trace::None);
+
+struct TxContext<'a> {
+    state: TxState,
+    tx: &'a mut [u8],
+    count: usize,  // index for transmit
+    end: usize,    // end of packet index
+    rot_irq: bool, // Track ROT_IRQ state
+}
+
+struct RxContext<'a> {
+    state: RxState,
+    rx: &'a mut [u8],
+    count: usize, // cursor for receive
+    end: usize,   // cursor for receive
+}
+
+enum EnqueueBuf<'a> {
+    Copy(&'a [u8]),
+    _TxBuf(usize),
+    Empty,
+}
+
+impl<'a> TxContext<'a> {
+    pub fn new(tx: &'a mut [u8]) -> Self {
+        Self {
+            state: TxState::Idle,
+            tx,
+            count: 0,
+            end: 0,
+            rot_irq: false,
+        }
+    }
+
+    // Enqueue a message for Tx to the SP.
+    // The message data may be zero-length (header only),
+    // copied from an external buffer, or the Tx buffer may already
+    // contain the data so only the length needs to be provided.
+    pub fn enqueue(&mut self, msgtype: MsgType, buf: EnqueueBuf) -> bool {
+        if self.state != TxState::Idle {
+            ringbuf_entry!(Trace::AlreadyEnqueuedMsg);
+            return false;
+        }
+        let len = match buf {
+            EnqueueBuf::Copy(buf) => buf.len(),
+            EnqueueBuf::_TxBuf(len) => len,
+            EnqueueBuf::Empty => 0,
+        };
+        if len > self.tx.len() {
+            ringbuf_entry!(Trace::EnqueueTooBig);
+            return false;
+        }
+
+        let mut msg = Msg::parse(&mut *self.tx).unwrap_lite();
+        msg.set_version();
+        msg.set_len(len);
+        msg.set_msgtype(msgtype);
+
+        if let EnqueueBuf::Copy(input) = buf {
```

Comment:
This would be more clear as an explicit `match`, so that we know that the other cases are deliberately ignored / don't need to copy any data.